### PR TITLE
catalog: Add shadow catalog implementation

### LINKED ISSUE #22481: Catalog: Add shadow implementation to compare Stash and Persist
### Feature request

In order to compare the Stash implementation of the `DurableCatalogState` to the Persist implementation, we should add a shadow implementation. The shadow implementation will internally have both a Stash and Persist implementation. On every method call, it will compare the results of both implementations.

On every write we might even want to compare snapshots of the catalog.

### PR DESCRIPTION
**Pull request by jkosh44**
This commit adds a shadow catalog implementation, which wraps both a
stash and a persist implementation. Each method call will dispatch the
call to both internal implementations and compare the results.

This commit also updates a handful of tests to use the shadow
implementation.

Resolves #22481

### Motivation
This PR adds a known-desirable feature.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

**Comment by jkosh44** [comment]
Just for the record, as of this comment, this PR seems to work as advertised. However, I'm seeing some pretty significant slowdown of Rust test.

When comparing the tests of the most recent build on main (https://buildkite.com/materialize/tests/builds/66772#018b62a6-a4aa-449b-8523-72e1db92ec52) and the most recent build for this PR (https://buildkite.com/materialize/tests/builds/66778#018b62cd-0f22-488d-b006-859518b47d8e), the following tests have a slowdown of larger than 1 second.


```
Slow down for	mz-environmentd::auth test_auth_base_disable_tls:                             61.37199999999999s
Slow down for	mz-environmentd::auth test_auth_admin_superuser:                              61.145999999999994s
Slow down for	mz-environmentd::auth test_auth_admin_non_superuser:                          60.632999999999996s
Slow down for	mz-environmentd::auth test_auth_admin_superuser_revoked:                      64.365s
Slow down for	mz-environmentd::auth test_auth_intermediate_ca:                              6.285s
Slow down for	mz-environmentd::auth test_auth_intermediate_ca_no_intermediary:              5.531000000000001s
Slow down for	mz-adapter catalog::tests::test_effective_search_path:                        16.082s
Slow down for	mz-adapter catalog::tests::test_minimal_qualification:                        8.117999999999999s
Slow down for	mz-adapter catalog::tests::test_normalized_create:                            15.611999999999998s
Slow down for	mz-adapter catalog::tests::test_compare_builtins_postgres:                    14.676000000000002s
Slow down for	mz-adapter catalog::tests::test_get_privileges:                               6.834s
Slow down for	mz-adapter catalog::tests::test_object_type:                                  15.203000000000001s
Slow down for	mz-adapter catalog::tests::test_pg_views_forbidden_types:                     14.152999999999999s
Slow down for	mz-environmentd::auth test_refresh_task_metrics:                              4.7509999999999994s
Slow down for	mz-adapter catalog::tests::test_catalog_revision:                             7.0070000000000014s
Slow down for	mz-persist-txn txns::tests::stress_correctness:                               8.241s
Slow down for	mz-stash tests::test_stash_fail_after_commit:                                 9.357000000000001s
Slow down for	mz-stash upgrade::v35_to_v36::tests::smoke_test_empty_flags:                  6.2379999999999995s
Slow down for	mz-stash upgrade::v35_to_v36::tests::smoke_test_existing_flags:               7.205s
Slow down for	mz-stash tests::test_stash_append:                                            6.323s
Slow down for	mz-stash tests::test_stash_savepoint:                                         9.138s
Slow down for	mz-stash tests::test_stash_stash:                                             59.547s
Slow down for	mz-stash upgrade::v39_to_v40::tests::smoke_test_audit_log_migration:          17.503s
Slow down for	mz-stash tests::test_stash_fence:                                             62.912s
Slow down for	mz-catalog::open test_debug_stash_is_initialized:                             48.403999999999996s
Slow down for	mz-stash upgrade::v38_to_v39::tests::smoke_test_role_vars_added:              2.8970000000000002s
Slow down for	mz-adapter catalog::tests::test_large_catalog_item:                           55.67600000000001s
Slow down for	mz-catalog::open test_stash_open_read_only:                                   46.694s
Slow down for	mz-adapter catalog::tests::test_smoketest_all_builtins:                       65.418s
Slow down for	mz-stash upgrade::v39_to_v40::tests::smoke_test_cluster_replica_migration:    59.36000000000001s
Slow down for	mz-catalog::open test_stash_get_deployment_generation:                        53.726s
Slow down for	mz-catalog::open test_debug_stash_open:                                       48.079s
Slow down for	mz-stash tests::test_stash_readonly:                                          1.0069999999999997s
Slow down for	mz-catalog::open test_debug_stash_get_deployment_generation:                  55.50600000000001s
Slow down for	mz-catalog::open test_stash_is_initialized:                                   52.975s
Slow down for	mz-catalog::open test_stash_open_savepoint:                                   52.065000000000005s
Slow down for	mz-catalog::open test_debug_stash_open_read_only:                             46.483999999999995s
Slow down for	mz-catalog::open test_debug_stash_open_savepoint:                             51.749s
Slow down for	mz-catalog::open test_stash_open:                                             50.31s
Slow down for	mz-environmentd::pgwire test_arrays:                                          3.7720000000000002s
Slow down for	mz-environmentd::pgwire test_conn_startup:                                    51.43899999999999s
Slow down for	mz-environmentd::pgwire test_conn_user:                                       34.659s
Slow down for	mz-environmentd::pgwire test_copy:                                            4.478s
Slow down for	mz-adapter::parameters test_parameter_type_inference:                         3.0919999999999987s
Slow down for	mz-environmentd::pgwire test_simple_query_no_hang:                            8.46s
Slow down for	mz-environmentd::pgwire test_record_types:                                    5.404999999999999s
Slow down for	mz-catalog::read-write test_stash_audit_logs:                                 1.452s
Slow down for	mz-catalog::read-write test_stash_catalog_content_version:                    2.708s
Slow down for	mz-catalog::read-write test_stash_timestamps:                                 8.246s
Slow down for	mz-catalog::read-write test_stash_system_privileges:                          6.417s
Slow down for	mz-catalog::read-write test_stash_system_items:                               5.947s
Slow down for	mz-adapter catalog::open::test_builtin_migration:                             35.346999999999994s
Slow down for	mz-balancerd::server test_balancer:                                           6.231s
Slow down for	mz-environmentd::server concurrent_cluster_drop:                              6.240999999999998s
Slow down for	mz-environmentd::server copy_from:                                            6.371999999999998s
Slow down for	mz-environmentd::server test_cancel_dataflow_removal:                         4.8759999999999994s
Slow down for	mz-environmentd::server test_cancel_insert_select:                            8.055s
Slow down for	mz-environmentd::server test_cancel_long_running_query:                       6.943000000000001s
Slow down for	mz-environmentd::server test_cancel_long_select:                              5.939s
Slow down for	mz-environmentd::server test_cancel_read_then_write:                          5.658999999999999s
Slow down for	mz-environmentd::server test_cancel_ws:                                       5.271000000000001s
Slow down for	mz-environmentd::server smoketest_webhook_source:                             5.628s
Slow down for	mz-environmentd::server test_closing_connection_for_insert_select:            4.026s
Slow down for	mz-environmentd::pgwire test_pgtest_mz:                                       35.59900000000002s
Slow down for	mz-environmentd::server test_closing_connection_for_long_select:              6.387000000000002s
Slow down for	mz-environmentd::server test_default_cluster_sizes:                           5.738000000000001s
Slow down for	mz-environmentd::server test_github_20262:                                    6.558999999999999s
Slow down for	mz-environmentd::server test_http_metrics:                                    5.971000000000002s
Slow down for	mz-environmentd::server test_http_options_param:                              5.999999999999998s
Slow down for	mz-environmentd::server test_internal_console_proxy:                          5.862000000000002s
Slow down for	mz-environmentd::server test_internal_http_auth:                              5.728s
Slow down for	mz-environmentd::server test_internal_ws_auth:                                6.139000000000001s
Slow down for	mz-environmentd::server test_http_sql:                                        13.460999999999999s
Slow down for	mz-environmentd::server test_invalid_webhook_body:                            5.915000000000001s
Slow down for	mz-environmentd::pgwire test_pgtest:                                          76.12299999999999s
Slow down for	mz-environmentd::server test_max_connections_on_all_interfaces:               5.764000000000001s
Slow down for	mz-environmentd::server test_leader_promotion_always_using_deploy_generation: 6.509999999999998s
Slow down for	mz-environmentd::server test_concurrent_id_reuse:                             4.803999999999988s
Slow down for	mz-environmentd::server test_leader_promotion:                                6.058s
Slow down for	mz-environmentd::server test_max_request_size:                                6.461s
Slow down for	mz-environmentd::server test_max_statement_batch_size:                        6.982000000000001s
Slow down for	mz-environmentd::server test_mz_system_user_admin:                            5.889000000000001s
Slow down for	mz-environmentd::server test_source_sink_size_required:                       6.3309999999999995s
Slow down for	mz-environmentd::server test_old_storage_usage_records_are_reaped_on_restart: 6.661999999999999s
Slow down for	mz-environmentd::server test_statement_logging_basic:                         3.655000000000001s
Slow down for	mz-environmentd::server test_persistence:                                     7.749000000000002s
Slow down for	mz-environmentd::server test_statement_logging_immediate:                     5.722999999999999s
Slow down for	mz-environmentd::server test_statement_logging_sampling:                      5.660999999999998s
Slow down for	mz-environmentd::server test_statement_logging_sampling_constrained:          6.079999999999998s
Slow down for	mz-environmentd::server test_statement_logging_unsampled_metrics:             6.443000000000001s
Slow down for	mz-environmentd::server test_statement_logging_subscribes:                    7.564999999999998s
Slow down for	mz-environmentd::server test_storage_usage_collection_interval:               8.113999999999997s
Slow down for	mz-environmentd::server test_storage_usage_collection_interval_timestamps:    5.859s
Slow down for	mz-environmentd::server test_webhook_duplicate_headers:                       5.968s
Slow down for	mz-environmentd::server test_storage_usage_doesnt_update_between_restarts:    2.5429999999999993s
Slow down for	mz-environmentd::server test_storage_usage_updates_between_restarts:          5.931999999999999s
Slow down for	mz-environmentd::server test_webhook_url_notice:                              5.904000000000002s
Slow down for	mz-environmentd::server test_ws_notifies_for_bad_options:                     5.932s
Slow down for	mz-environmentd::server test_ws_passes_options:                               6.277999999999999s
Slow down for	mz-environmentd::server webhook_too_large_request:                            6.371999999999998s
Slow down for	mz-environmentd::server webhook_concurrency_limit:                            6.6320000000000014s
Slow down for	mz-environmentd::server test_statement_logging_persistence:                   6.337999999999994s
Slow down for	mz-environmentd::server webhook_concurrent_actions:                           5.969999999999999s
Slow down for	mz-environmentd::sql test_alter_system_invalid_param:                         6.045s
Slow down for	mz-environmentd::sql test_auto_run_on_introspection_feature_disabled:         6.099s
Slow down for	mz-environmentd::sql test_auto_run_on_introspection_feature_enabled:          5.65s
Slow down for	mz-environmentd::sql test_auto_run_on_introspection_per_replica_relations:    5.668000000000001s
Slow down for	mz-environmentd::sql test_dont_drop_sinks_twice:                              6.699000000000002s
Slow down for	mz-environmentd::sql test_drop_connection_race:                               5.959000000000001s
Slow down for	mz-environmentd::sql test_emit_timestamp_notice:                              6.289999999999999s
Slow down for	mz-environmentd::sql test_emit_tracing_notice:                                6.480999999999998s
Slow down for	mz-environmentd::sql test_concurrent_writes:                                  5.592999999999996s
Slow down for	mz-environmentd::sql test_empty_subscribe_error:                              5.489000000000001s
Slow down for	mz-environmentd::sql test_empty_subscribe_notice:                             5.961s
Slow down for	mz-environmentd::sql test_coord_startup_blocking:                             8.809999999999995s
Slow down for	mz-environmentd::sql test_explain_timestamp_json:                             6.116s
Slow down for	mz-environmentd::sql test_explain_timestamp_table:                            6.3100000000000005s
Slow down for	mz-environmentd::sql test_github_12951:                                       6.471s
Slow down for	mz-environmentd::sql test_idle_in_transaction_session_timeout:                6.009s
Slow down for	mz-environmentd::sql test_internal_ports:                                     1.980999999999998s
Slow down for	mz-environmentd::sql test_load_generator:                                     1.4179999999999993s
Slow down for	mz-environmentd::sql test_max_connections:                                    5.411000000000001s
Slow down for	mz-environmentd::sql test_no_block:                                           5.959s
Slow down for	mz-environmentd::sql test_mz_sessions:                                        5.956000000000003s
Slow down for	mz-environmentd::sql test_params:                                             5.780999999999999s
Slow down for	mz-environmentd::sql test_peek_on_dropped_cluster:                            5.5760000000000005s
Slow down for	mz-environmentd::sql test_peek_on_dropped_indexed_view:                       5.468000000000002s
Slow down for	mz-environmentd::sql test_pg_cancel_backend:                                  6.042000000000002s
Slow down for	mz-environmentd::sql test_pg_cancel_dropped_role:                             6.158000000000001s
Slow down for	mz-environmentd::sql test_read_then_write_serializability:                    6.2429999999999986s
Slow down for	mz-environmentd::sql test_subcribe_continuous_progress:                       6.376999999999999s
Slow down for	mz-environmentd::sql test_subscribe_consolidation:                            6.577999999999999s
Slow down for	mz-environmentd::sql test_subscribe_basic:                                    6.4929999999999986s
Slow down for	mz-environmentd::sql test_subscribe_empty_upper_frontier:                     5.920999999999999s
Slow down for	mz-environmentd::sql test_subscribe_fetch_wait:                               7.161s
Slow down for	mz-environmentd::sql test_subscribe_fetch_timeout:                            6.596s
Slow down for	mz-environmentd::sql test_subscribe_negative_diffs:                           6.558999999999999s
Slow down for	mz-environmentd::sql test_subscribe_on_dropped_source:                        7.113999999999999s
Slow down for	mz-environmentd::sql test_subscribe_outlive_cluster:                          7.058999999999999s
Slow down for	mz-environmentd::sql test_subscribe_progress_non_nullable_columns:            6.468000000000002s
Slow down for	mz-environmentd::sql test_subscribe_shutdown:                                 5.827000000000002s
Slow down for	mz-environmentd::sql test_subscribe_table_rw_timestamps:                      5.863999999999999s
Slow down for	mz-environmentd::sql test_temporary_views:                                    5.996s
Slow down for	mz-environmentd::sql test_time:                                               4.561000000000002s
Slow down for	mz-environmentd::sql test_timelines_persist_after_failed_transaction:         5.013s
Slow down for	mz-environmentd::sql test_timeline_read_holds:                                6.362s
Slow down for	mz-environmentd::sql test_utilization_hold:                                   6.369s
Slow down for	mz-environmentd::sql test_transactional_explain_timestamps:                   6.172999999999998s
Slow down for	mz-environmentd::sql test_timestamp_recovery:                                 5.303000000000001s
Slow down for	mz-adapter::sql datadriven:                                                   5.819s
Slow down for	mz-environmentd::sql test_subscribe_progress:                                 47.443s
```

**Comment by jkosh44** [comment]
Running things in parallel looks much better, but still not great:

```
Slow down for	mz-environmentd::auth test_auth_base_disable_tls:                             2.061s;               original time: 22.573s
Slow down for	mz-environmentd::auth test_auth_admin_superuser:                              1.9830000000000005s;  original time: 23.058s
Slow down for	mz-environmentd::auth test_auth_admin_non_superuser:                          1.480999999999998s;   original time: 23.559s
Slow down for	mz-environmentd::auth test_auth_admin_superuser_revoked:                      45.897000000000006s;  original time: 36.872s
Slow down for	mz-environmentd::auth test_auth_intermediate_ca:                              5.670000000000002s;   original time: 11.822s
Slow down for	mz-environmentd::auth test_auth_intermediate_ca_no_intermediary:              5.372999999999999s;   original time: 11.844s
Slow down for	mz-compute-types plan::tests::plan_protobuf_roundtrip:                        2.0530000000000004s;  original time: 2.299s
Slow down for	mz-adapter catalog::tests::test_effective_search_path:                        6.645s;               original time: 8.887s
Slow down for	mz-adapter catalog::tests::test_minimal_qualification:                        6.036999999999999s;   original time: 9.003s
Slow down for	mz-adapter catalog::tests::test_normalized_create:                            6.603999999999999s;   original time: 9.073s
Slow down for	mz-adapter catalog::tests::test_compare_builtins_postgres:                    4.603s;               original time: 10.026s
Slow down for	mz-adapter catalog::tests::test_get_privileges:                               4.122999999999999s;   original time: 10.134s
Slow down for	mz-compute-types plan::join::tests::join_plan_protobuf_roundtrip:             4.371s;               original time: 7.292s
Slow down for	mz-adapter catalog::tests::test_object_type:                                  5.833s;               original time: 9.181s
Slow down for	mz-adapter catalog::tests::test_pg_views_forbidden_types:                     4.825999999999999s;   original time: 9.794s
Slow down for	mz-environmentd::auth test_refresh_task_metrics:                              9.531s;               original time: 14.347s
Slow down for	mz-adapter catalog::tests::test_catalog_revision:                             5.463999999999999s;   original time: 18.805s
Slow down for	mz-stash tests::test_stash_fail_after_commit:                                 3.5589999999999997s;  original time: 2.103s
Slow down for	mz-stash upgrade::v39_to_v40::tests::smoke_test_audit_log_migration:          7.914s;               original time: 2.249s
Slow down for	mz-adapter catalog::tests::test_large_catalog_item:                           2.217000000000006s;   original time: 39.733s
Slow down for	mz-catalog::open test_stash_open_read_only:                                   3.490000000000001s;   original time: 5.153s
Slow down for	mz-adapter catalog::tests::test_smoketest_all_builtins:                       4.640999999999998s;   original time: 38.356s
Slow down for	mz-environmentd::pgwire test_arrays:                                          5.825999999999999s;   original time: 14.819s
Slow down for	mz-environmentd::pgwire test_conn_startup:                                    4.7010000000000005s;  original time: 14.393s
Slow down for	mz-environmentd::pgwire test_conn_user:                                       5.3039999999999985s;  original time: 12.375s
Slow down for	mz-environmentd::pgwire test_copy:                                            4.917s;               original time: 12.345s
Slow down for	mz-adapter::parameters test_parameter_type_inference:                         2.722999999999999s;   original time: 56.072s
Slow down for	mz-environmentd::pgwire test_simple_query_no_hang:                            5.254999999999999s;   original time: 10.829s
Slow down for	mz-environmentd::pgwire test_record_types:                                    5.679s;               original time: 11.812s
Slow down for	mz-catalog::read-write test_stash_comments:                                   3.666s;               original time: 1.363s
Slow down for	mz-catalog::read-write test_stash_audit_logs:                                 3.2119999999999997s;  original time: 1.896s
Slow down for	mz-catalog::read-write test_stash_introspection_source_indexes:               1.625s;               original time: 2.816s
Slow down for	mz-catalog::read-write test_stash_default_privileges:                         1.484s;               original time: 3.107s
Slow down for	mz-catalog::read-write test_stash_system_privileges:                          1.9520000000000004s;  original time: 2.272s
Slow down for	mz-adapter catalog::open::test_builtin_migration:                             45.616s;              original time: 120.986s
Slow down for	mz-balancerd::server test_balancer:                                           5.484999999999998s;   original time: 12.345s
Slow down for	mz-environmentd::server concurrent_cluster_drop:                              6.017999999999999s;   original time: 12.65s
Slow down for	mz-environmentd::server copy_from:                                            5.2559999999999985s;  original time: 12.249s
Slow down for	mz-environmentd::server test_cancel_dataflow_removal:                         5.42s;                original time: 15.409s
Slow down for	mz-environmentd::server test_cancel_insert_select:                            6.300999999999998s;   original time: 13.872s
Slow down for	mz-environmentd::server test_cancel_long_running_query:                       5.097000000000001s;   original time: 12.875s
Slow down for	mz-environmentd::server test_cancel_long_select:                              4.0219999999999985s;  original time: 16.527s
Slow down for	mz-environmentd::server test_cancel_read_then_write:                          5.403999999999996s;   original time: 31.954s
Slow down for	mz-environmentd::server test_cancel_ws:                                       4.917000000000002s;   original time: 12.049s
Slow down for	mz-environmentd::server smoketest_webhook_source:                             5.509s;               original time: 112.601s
Slow down for	mz-environmentd::server test_closing_connection_for_insert_select:            5.154999999999998s;   original time: 17.434s
Slow down for	mz-environmentd::pgwire test_pgtest_mz:                                       71.72000000000003s;   original time: 199.176s
Slow down for	mz-environmentd::server test_closing_connection_for_long_select:              7.787000000000001s;   original time: 13.876s
Slow down for	mz-environmentd::server test_default_cluster_sizes:                           11.416000000000002s;  original time: 11.126s
Slow down for	mz-environmentd::server test_github_20262:                                    7.095999999999998s;   original time: 11.749s
Slow down for	mz-environmentd::server test_http_metrics:                                    5.017000000000001s;   original time: 11.449s
Slow down for	mz-environmentd::server test_http_options_param:                              7.286s;               original time: 11.627s
Slow down for	mz-environmentd::server test_internal_console_proxy:                          5.969999999999999s;   original time: 11.177s
Slow down for	mz-environmentd::server test_internal_http_auth:                              4.735000000000001s;   original time: 12.245s
Slow down for	mz-environmentd::server test_internal_ws_auth:                                10.299000000000001s;  original time: 11.048s
Slow down for	mz-environmentd::server test_http_sql:                                        12.052999999999997s;  original time: 46.963s
Slow down for	mz-environmentd::server test_invalid_webhook_body:                            7.6800000000000015s;  original time: 12.859s
Slow down for	mz-environmentd::pgwire test_pgtest:                                          98.935s;              original time: 280.419s
Slow down for	mz-environmentd::server test_max_connections_on_all_interfaces:               6.511000000000001s;   original time: 12.144s
Slow down for	mz-environmentd::server test_leader_promotion_always_using_deploy_generation: 9.111999999999998s;   original time: 21.667s
Slow down for	mz-environmentd::server test_concurrent_id_reuse:                             7.541999999999987s;   original time: 99.537s
Slow down for	mz-environmentd::server test_leader_promotion:                                4.244s;               original time: 32.506s
Slow down for	mz-environmentd::server test_max_request_size:                                4.807999999999998s;   original time: 11.752s
Slow down for	mz-environmentd::server test_max_statement_batch_size:                        5.546000000000001s;   original time: 10.996s
Slow down for	mz-environmentd::server test_mz_system_user_admin:                            5.254s;               original time: 10.673s
Slow down for	mz-environmentd::server test_source_sink_size_required:                       7.535s;               original time: 12.212s
Slow down for	mz-environmentd::server test_old_storage_usage_records_are_reaped_on_restart: 11.776000000000003s;  original time: 22.375s
Slow down for	mz-environmentd::server test_statement_logging_basic:                         5.361000000000001s;   original time: 17.938s
Slow down for	mz-environmentd::server test_persistence:                                     9.715000000000003s;   original time: 23.58s
Slow down for	mz-environmentd::server test_statement_logging_immediate:                     5.4339999999999975s;  original time: 19.167s
Slow down for	mz-environmentd::server test_statement_logging_sampling:                      5.239000000000001s;   original time: 17.018s
Slow down for	mz-environmentd::server test_statement_logging_sampling_constrained:          7.573999999999998s;   original time: 17.166s
Slow down for	mz-environmentd::server test_statement_logging_unsampled_metrics:             5.2669999999999995s;  original time: 10.738s
Slow down for	mz-environmentd::server test_statement_logging_subscribes:                    9.276s;               original time: 18.411s
Slow down for	mz-environmentd::server test_storage_usage_collection_interval:               12.084s;              original time: 17.952s
Slow down for	mz-environmentd::server test_storage_usage_collection_interval_timestamps:    5.205s;               original time: 12.039s
Slow down for	mz-environmentd::server test_webhook_duplicate_headers:                       8.964000000000002s;   original time: 11.86s
Slow down for	mz-environmentd::server test_storage_usage_doesnt_update_between_restarts:    1.7310000000000016s;  original time: 30.793s
Slow down for	mz-environmentd::server test_storage_usage_updates_between_restarts:          5.042999999999999s;   original time: 26.779s
Slow down for	mz-environmentd::server test_webhook_url_notice:                              5.203000000000001s;   original time: 15.158s
Slow down for	mz-environmentd::server test_ws_notifies_for_bad_options:                     5.51s;                original time: 10.524s
Slow down for	mz-environmentd::server test_ws_passes_options:                               11.347999999999999s;  original time: 10.375s
Slow down for	mz-environmentd::server webhook_too_large_request:                            5.729000000000001s;   original time: 11.711s
Slow down for	mz-environmentd::server webhook_concurrency_limit:                            5.809000000000001s;   original time: 17.203s
Slow down for	mz-environmentd::server test_statement_logging_persistence:                   6.6499999999999915s;  original time: 92.26s
Slow down for	mz-environmentd::server webhook_concurrent_actions:                           5.777000000000001s;   original time: 20.267s
Slow down for	mz-environmentd::sql test_alter_system_invalid_param:                         6.929s;               original time: 10.931s
Slow down for	mz-environmentd::sql test_auto_run_on_introspection_feature_disabled:         5.652000000000001s;   original time: 13.296s
Slow down for	mz-environmentd::sql test_auto_run_on_introspection_feature_enabled:          15.571s;              original time: 12.507s
Slow down for	mz-environmentd::sql test_auto_run_on_introspection_per_replica_relations:    10.548s;              original time: 12.374s
Slow down for	mz-environmentd::sql test_dont_drop_sinks_twice:                              6.823s;               original time: 13.439s
Slow down for	mz-environmentd::sql test_drop_connection_race:                               5.5329999999999995s;  original time: 12.516s
Slow down for	mz-environmentd::sql test_emit_timestamp_notice:                              5.109999999999999s;   original time: 12.073s
Slow down for	mz-environmentd::sql test_emit_tracing_notice:                                5.315999999999999s;   original time: 12.067s
Slow down for	mz-environmentd::sql test_concurrent_writes:                                  6.212999999999994s;   original time: 41.892s
Slow down for	mz-environmentd::sql test_empty_subscribe_error:                              10.91s;               original time: 11.329s
Slow down for	mz-environmentd::sql test_empty_subscribe_notice:                             5.272s;               original time: 11.374s
Slow down for	mz-environmentd::sql test_coord_startup_blocking:                             6.705999999999996s;   original time: 46.215s
Slow down for	mz-environmentd::sql test_explain_timestamp_json:                             5.736999999999998s;   original time: 10.771s
Slow down for	mz-environmentd::sql test_explain_timestamp_table:                            22.866s;              original time: 10.774s
Slow down for	mz-environmentd::sql test_github_12951:                                       37.902s;              original time: 13.361s
Slow down for	mz-environmentd::sql test_idle_in_transaction_session_timeout:                35.743s;              original time: 13.171s
Slow down for	mz-environmentd::sql test_internal_ports:                                     30.091s;              original time: 15.387s
Slow down for	mz-environmentd::sql test_internal_users:                                     5.621000000000002s;   original time: 21.008s
Slow down for	mz-environmentd::sql test_max_connections:                                    4.460000000000001s;   original time: 13.541s
Slow down for	mz-environmentd::sql test_no_block:                                           5.330000000000002s;   original time: 11.703s
Slow down for	mz-environmentd::sql test_mz_sessions:                                        5.261000000000003s;   original time: 18.487s
Slow down for	mz-environmentd::sql test_params:                                             5.484999999999999s;   original time: 11.425s
Slow down for	mz-environmentd::sql test_peek_on_dropped_cluster:                            5.373999999999999s;   original time: 13.16s
Slow down for	mz-environmentd::sql test_peek_on_dropped_indexed_view:                       4.711s;               original time: 15.806s
Slow down for	mz-environmentd::sql test_pg_cancel_backend:                                  5.371000000000002s;   original time: 12.363s
Slow down for	mz-environmentd::sql test_pg_cancel_dropped_role:                             10.155999999999999s;  original time: 12.026s
Slow down for	mz-environmentd::sql test_read_then_write_serializability:                    7.327999999999999s;   original time: 13.134s
Slow down for	mz-environmentd::sql test_subcribe_continuous_progress:                       5.687999999999999s;   original time: 13.396s
Slow down for	mz-environmentd::sql test_subscribe_consolidation:                            6.324000000000002s;   original time: 11.378s
Slow down for	mz-environmentd::sql test_subscribe_basic:                                    5.774999999999999s;   original time: 15.067s
Slow down for	mz-environmentd::sql test_subscribe_empty_upper_frontier:                     5.899999999999999s;   original time: 11.84s
Slow down for	mz-environmentd::sql test_subscribe_fetch_wait:                               6.090000000000002s;   original time: 11.769s
Slow down for	mz-environmentd::sql test_subscribe_fetch_timeout:                            5.233999999999998s;   original time: 14.242s
Slow down for	mz-environmentd::sql test_subscribe_negative_diffs:                           8.747000000000002s;   original time: 11.949s
Slow down for	mz-environmentd::sql test_subscribe_on_dropped_source:                        9.132s;               original time: 14.304s
Slow down for	mz-environmentd::sql test_subscribe_outlive_cluster:                          5.929s;               original time: 12.804s
Slow down for	mz-environmentd::sql test_subscribe_progress_non_nullable_columns:            5.291000000000002s;   original time: 11.524s
Slow down for	mz-environmentd::sql test_subscribe_shutdown:                                 5.2890000000000015s;  original time: 12.404s
Slow down for	mz-environmentd::sql test_subscribe_table_rw_timestamps:                      5.415000000000001s;   original time: 12.039s
Slow down for	mz-environmentd::sql test_temporary_views:                                    4.9129999999999985s;  original time: 12.22s
Slow down for	mz-environmentd::sql test_time:                                               4.634000000000002s;   original time: 12.53s
Slow down for	mz-environmentd::sql test_timelines_persist_after_failed_transaction:         5.296999999999999s;   original time: 13.473s
Slow down for	mz-environmentd::sql test_timeline_read_holds:                                5.505000000000001s;   original time: 14.99s
Slow down for	mz-environmentd::sql test_utilization_hold:                                   8.764s;               original time: 11.517s
Slow down for	mz-environmentd::sql test_transactional_explain_timestamps:                   5.09s;                original time: 13.887s
Slow down for	mz-environmentd::sql test_timestamp_recovery:                                 4.652000000000001s;   original time: 22.836s
Slow down for	mz-adapter::sql datadriven:                                                   12.042000000000002s;  original time: 7.985s
Slow down for	mz-environmentd::sql test_subscribe_progress:                                 42.79899999999999s;   original time: 94.158s
```

**Comment by jkosh44** [comment]
Note there doesn't seem to be any noticeable slowdown in SLT tests.

**Comment by jkosh44** [comment]
@MaterializeInc/testing just confirming, those nightly testdrive failures are not expected right? It's a bit strange since I don't think this PR touches testdrive.

**Comment by def-** [comment]
I haven't seen the failure in nightly: https://buildkite.com/materialize/nightlies/builds/4944

Is it possible that the database cleanup inbetween tests in testdrive is broken by this change?

**Review by philip-stoev — APPROVED** [review]
Nightly failures unrelated.

**Review by ParkMyCar — APPROVED** [review]

File: src/catalog/src/durable.rs [hunk 1]
```diff
@@ -265,6 +265,18 @@ pub async fn persist_backed_catalog_state(
     PersistHandle::new(persist_client, organization_id).await
 }
 
+/// Creates an openable durable catalog state implemented using both the stash and persist, that
+/// compares the results.
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/catalog/src/durable.rs]: Done.

File: src/ore/src/assert.rs [hunk 2]
```diff
@@ -107,6 +107,27 @@ macro_rules! soft_assert_or_log {
...
+    }};
+    ($left:expr, $right:expr, $($arg:tt)+) => {{
+        if $crate::assert::SOFT_ASSERTIONS.load(::std::sync::atomic::Ordering::Relaxed) {
+            assert_eq!($left, $right, $($arg)+);
+        } else if $left != $right{
+            ::tracing::error!("assertion {:?} == {:?} failed: {}", $left, $right, format!($($arg)+))
+        }
+    }};
+}
+
```
Comment (jkosh44) [inline src/ore/src/assert.rs:130]: @ParkMyCar Can you take a look at this? `soft_assert_or_log` didn't really match the semantics that I was looking for.

Comment (ParkMyCar) [reply src/ore/src/assert.rs:130]: Looks good to me, thanks for adding this!

**Review by ParkMyCar — APPROVED** [review]

**Pull request merged**